sdl2promela/translator.py

```diff
@@ -296,36 +386,498 @@ def __generate_statement(
+            )
+        length_field = (
+            MemberAccessBuilder()
+            .withUtypeReference(__generate_variable_name(sdl_model, left, True))
+            .withMember(VariableReferenceBuilder("length").build())
+            .build()
+        )
+
+        statements: List[promelamodel.Statement] = []
+
+        if int(finalType.Min) != int(finalType.Max):
+            statements.append(
+                AssignmentBuilder()
+                .withTarget(length_field)
+                .withSource(promelamodel.IntegerValue(length))
+                .build()
+            )
+        for index in range(int(finalType.Max)):
+            if index < length:
+                value = promelamodel.IntegerValue(right.elements[index])
+            else:
+                value = promelamodel.IntegerValue(0)
+            element = (
+                ArrayAccessBuilder()
+                .withArray(__generate_variable_name(sdl_model, left, True))
+                .withIndex(promelamodel.IntegerValue(index))
+                .build()
+            )
+            statements.append(
+                AssignmentBuilder().withTarget(element).withSource(value).build()
+            )
+
+        return statements
+    elif finalType.kind == "BitStringType":
+        raise NotImplementedError("Assignment to BIT STRING is not supported")
+    else:
+        raise Exception(f"Unsupported assignment: {finalType.kind} EmptyString")
+
+
+@dispatch(
+    sdlmodel.Model,
+    (
+        sdlmodel.VariableReference,
+        sdlmodel.ArrayAccess,
+        sdlmodel.MemberAccess,
+    ),
+    sdlmodel.Sequence,
+    type,
+)
+def __generate_assignment(
+    sdl_model: sdlmodel.Model,
+    left: Union[
+        sdlmodel.VariableReference, sdlmodel.ArrayAccess, sdlmodel.MemberAccess
+    ],
+    right: sdlmodel.Sequence,
+    left_type: type,
+) -> List[promelamodel.Statement]:
+    finalType = resolve_asn1_type(sdl_model.types, left_type)
+    if finalType.kind != "SequenceType":
+        raise Exception(
+            f"Invalid assignment: {finalType.CName} does not accept SEQUENCE value"
+        )
+
+    statements: List[promelamodel.Statement] = []
+
+    for name, dataType in finalType.Children.items():
+        is_optional = dataType.Optional == "True"
+        if not is_optional and name not in right.elements:
+            raise Exception(
+                f"Invalid assignment to {finalType.CName}: missing required member {name}"
+            )
+
+        if name in right.elements:
+            statements.extend(
+                __generate_assignment(
+                    sdl_model,
+                    sdlmodel.MemberAccess(left, sdlmodel.VariableReference(name)),
+                    right.elements[name],
+                    dataType.type,
+                )
+            )
+
+        if is_optional:
+            if name in right.elements:
+                value = promelamodel.IntegerValue(1)
```

Comment:
Consider value = 1 if name in right.elements else 0

Reply:
Done.